### ISSUE TITLE
[HOLD for payment 2023-08-28] [HOLD] [$1000]  Web - Clicking on the browser back button takes user from chat page to assign task page

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#23796]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open a chat
2. Click on the plus icon
3. Click on assign task
4. Give a title and click next
5. Confirm the task
6. From the chat screen click the browser back button

## Expected Result:
App should not open the assign task side panel

## Actual Result:
App takes user to the assign task screen

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.47-2
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

**Expensify/Expensify Issue URL:**
**Issue reported by:** @Nathan-Mulugeta
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1690536196870419

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01dd04525ab1095d86</li>
        <li>Upwork Job ID: 1685016837327048704</li>
        2023-07-28<li>Automatic offers: </li>
<ul>
<li>daordonez11 | Contributor | 25816261</li>
<li>Nathan-Mulugeta | Reporter | 25816263</li>
</ul></li>
    </ul>
</details>

**Comment by daordonez11** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Pressing the back button sends to task detail instead of parent report

### What is the root cause of that problem?

The `dismissModal` method in `Nagivation.js`  is using the type REPLACE hence in this special case the parent is replaced with the child so going back does not send to the parent report.

### What changes do you think we should make in order to solve the problem?

There is 2 ways of solving it:
1. I removed the line and normal navigate type is working as expected. Tested in 4 different modals and was working fine.
2. Create an algorithm that validates if targetReport is a child of topmostReport and if it is only remove REPLACE in this case something like:
```
if(!targetReport.isChild(getTopmostReportId(rootState)))
{
    action.type = 'REPLACE';
}
```

### What alternative solutions did you explore? (Optional)
If the action is executed correctly remove previous steps from navigation stack but seems too complex code for something so simple.

**Comment by mountiny** [comment]
Stepping in since its related to navigation

@daordonez11 I think your proposal looks good, but I would go with the option 2. I think we needed the replace there for  some other reasons

**Comment by mountiny** [comment]
@adamgrzybowski @WoLewicki do you remember why we had the replace there? I though it was for this exact reason so I am confused that the dismis modal does not work here

**Comment by daordonez11** [comment]
Got it @mountiny will test the cases! There are 33 usages of dismissModal so probably step 2 is the best way to go, will check if there is any util to validate if it is a child of the report and will test deeply the scenarios. If any of you guys remember the origin of why replace is there would be even better to narrow the validation, specially @WoLewicki  I think the first appearance of it is in the commit: d2b8e2 feat: first look on RHP Navigation created by Wojciech.

**Comment by WoLewicki** [comment]
Hmm this topic consists of two parts.

First is the `react-navigation` state. Here in all those RHP flows such as `Request money` or `Assign task`, we want the RHP to disappear after ending the flow and also we want to navigate to the new report immediately after it too. Since RHP is at at the root of the navigation stack, and new report is also there, we just replace the RHP with new report, that's why this code is there and should work correctly (the RHP disappears from navigation tree and new report appears).

Second thing is browser history. All those flows worked with browser buttons as long as they were a single url. It was done with the `steps` concept, where we navigated between the parts of the RHP flow with simple animations, but we did not change the url in different steps. I created this solution when those flows looked like this. It was just fine then since when calling `replace` action from `react-navigation`, we also call `replace` in browser stack, and since those flows were a single url, we just replaced the last url with the new one, making the back button lead to the previous report.

It changed when those flows started to be different urls. While the `react-navigation` state still works correctly, the browser history "remembers" all the previous urls, so we just replace the last page of those flows, and clicking back button leads us to the previous page of the flow. 

We already found a solution to drop all the entries of history associated with the nested stacks when they are dropped in the `react-navigation's` browser API integration and @adamgrzybowski should post a PR with it soon. While it will work in our case, we are unsure if it is the behavior most users of `react-navigation` library would expect, since browser history just follows the urls as the source of truth, so it might be tricky to have it merged upstream imo.

Is it clear enough? Or should I explain something better?

Another thing is that I am not sure why simply navigating in such case works correctly. @daordonez11 could you check if `RHP` really disappears from the navigation stack if you do not drop it explicitly? Or am I missing something?

**Comment by mountiny** [comment]
You are right, I have missed that this flow got refactored since then, great explanation, thank you very much

**Comment by daordonez11** [comment]
Thanks for the detailed explanation! @WoLewicki ita clear why it worked and why it's failing in this case! Let me implement a log to check the status after navigating. To be clear it works because usually dismissing the modal sends back to the same report in that case navigating or replacing does the same! Navigate works because it finds the report in the stack and goes back, and replace just sends directly. Let me do some tests. @mountiny  didn't raise PR because was waiting for some context to build the validation, also couldn't find an existing util to validate if one report is parent of the other.

**Comment by WoLewicki** [comment]
> Navigate works because it finds the report in the stack and goes back, and replace just sends directly.

I am afraid we cannot go with such solution since it can drop many reports on the way and we agreed that we always want to push new ones, not navigate back to the old ones after going to them again.

**Comment by daordonez11** [comment]
Noted thanks @WoLewicki then will make sure Navigate is only used if it is a child report and will validate status of the stack then! Thanks for the help. Also if any of you know a current util that does this would be awesome if not I can build it.

**Comment by WoLewicki** [comment]
https://github.com/Expensify/App/blob/ec1ccf8fd058299a94c6694553c4508dd62981ed/src/libs/Navigation/Navigation.js#L166 this check should handle the case where we should just go back since we were already in the target report. I think the logic here is alright, the problem is only with the browser history. Or am I missing the thing you are asking for @daordonez11 ?

**Comment by daordonez11** [comment]
Yeah that one validates if it is the same report! I'm looking for something that validates if targetReportId is a child of the topmost report! The rest is clear! Just to include an specific validation to use Navigate instead of Replace in that specific case.

**Comment by gadhiyamanan** [comment]
@mountiny i think this is dupe of https://github.com/Expensify/App/issues/21356

**Comment by WoLewicki** [comment]
Yeah, and my comment here is detailed explanation of this one: https://github.com/Expensify/App/issues/21356#issuecomment-1633930925 😅

**Comment by WoLewicki** [comment]
> Just to include an specific validation to use Navigate instead of Replace in that specific case.

Could you tell where this case would be needed? I believe there is a `parentReportID` prop or something close to it in the props of the thread, but I am not sure.

**Comment by daordonez11** [comment]
Hey guys @WoLewicki @mountiny  A first approach on ensuring that "navigate" is only used on this specific scenario https://github.com/Expensify/App/pull/23974 let me know what you think about this solution. I have gathered information regarding how to validate if the target report is a child. Still, there is no existing util or easy way of doing it, you need access to the report action, which is outside of the scope of "navigation" That's why I decided to use an extending param.

**Comment by daordonez11** [comment]
PS @WoLewicki  Tested and efectively the stack is cleaned using this approach, there is nothing regarding tasks after the navigation is executed. Let me know if it makes sense so I can test in all platforms.

**Comment by mountiny** [comment]
Assigning Wojti instead of Daniel as this will require more complex fix

**Comment by mountiny** [comment]
This is on HOLD for react navigation fix so that can take long, making it monthly

**Comment by adamgrzybowski** [comment]
Hey, this should be fixed by this [PR](https://github.com/Expensify/App/pull/24165)

**Comment by mountiny** [comment]
@michaelhaxhiu could you please confirm this issue is resolved in staging? If yes, then we can close it

**Comment by michaelhaxhiu** [comment]
The bug is resolved!

I paid the bug reporter just now @Nathan-Mulugeta 

Special shout out to @daordonez11 for his preliminary efforts in resolving this, it's much appreciated and hope to see you involved in more jobs in the future. 🙏

**Comment by eVoloshchak** [comment]
>  any context on why this old issue just come up for C+ payment in NewDot now?

@trjExpensify, there was a recent addition to the rules [here](https://expensify.slack.com/archives/C02NK2DQWUX/p1767824013951259)

> Request payment in the same month the payment summary is posted. If that is not possible, request it as soon as possible after month-end.

So I wanted to make sure I requested everything in the past, went through all of the PRs I have ever reviewed, and noticed I never requested a payment for reviewing https://github.com/Expensify/App/pull/24165

> The summary from the BZ here is suggesting only the bug reporting payment was due.

I did review https://github.com/Expensify/App/pull/24165

**Comment by trjExpensify** [comment]
Okay, I think @mallenexpensify is looking at these centrally. I'll leave it on hold pending further opinion. 👍

**Comment by mallenexpensify** [comment]
> I have the request on hold.

@trjExpensify I'm going to pay on the other, main issue. Please cancel/reject the payment request from @eVoloshchak here. Thx.
